File: src/gestorAplicacion/usuario/Usuario.java
package gestorAplicacion.usuario;
import java.io.Serializable;
import java.util.ArrayList;

import gestorAplicacion.administracion.Materia;


public abstract class Usuario implements Serializable{
    private static final long serialVersionUID = 1L;
    protected long id;
    protected String nombre;
    protected String tipo;
    protected String pw;
    protected String facultad;
    protected static ArrayList<Usuario> usuariosTotales = new ArrayList<Usuario>();

    public Usuario(long id, String nombre, String facultad) {
        this.id = id;
        this.nombre = nombre;
        this.facultad = facultad;
        usuariosTotales.add(this);
    }

    public Usuario(long id, String nombre, String pw, String facultad){
        this.id = id;
        this.nombre = nombre;
        this.pw = pw;
        this.facultad = facultad;
        usuariosTotales.add(this);
    }

    public abstract String toString();

    public static String mostrarUsuarios() {
    	String retorno = "";
    	int i = 1;
    	for(Usuario usuario:usuariosTotales) {
    		retorno += (i++)+". "+usuario.nombre+", id: "+usuario.id+"\n";
    	}
    	return retorno;
    }

    public boolean comprobacionFacultad(Usuario usuario){
        String facultad1 = this.getFacultad().toLowerCase();
        String facultad2 = usuario.getFacultad().toLowerCase();
        if (facultad1.equals(facultad2)){
            return true;
        }
        return false;
    }

    public void desmatricularDelSistema(Usuario usuario){

        for (Usuario u: Usuario.getUsuariosTotales()){

            if (usuario.equals(u)){
                    Usuario.getUsuariosTotales().remove(usuario);
                    break;
                }
        }
    }

    public void eliminarMateria(Materia materia){
    	Materia.getMateriasTotales().remove(materia);
    }

    public void agregarMateria(String nombre, int codigo, String descripcion,int creditos, String facultad, ArrayList<Materia> prerrequisitos){
    	Materia nMateria = new Materia(nombre, codigo, descripcion, creditos, facultad, prerrequisitos);
    	Materia.getMateriasTotales().add(nMateria);
    }

    public String getTipo() {
		return tipo;
	}

	public void setTipo(String tipo) {
		this.tipo = tipo;
	}

	public long getId() {
        return id;
    }
    public void setId(long id) {
        this.id = id;
    }


    public String getNombre() {
        return nombre;
    }

    public void setNombre(String nombre) {
        this.nombre = nombre;
    }

    public String getPw() {
        return pw;
    }
    public void setPw(String pw) {
        this.pw = pw;
    }
    
    public String getFacultad() {
        return facultad;
    }

    public void setFacultad(String facultad) {
        this.facultad = facultad;
    }

    public static ArrayList<Usuario> getUsuariosTotales() {
        return Usuario.usuariosTotales;
    }

    public static void setUsuariosTotales(ArrayList<Usuario> usuariosTotales) {
        Usuario.usuariosTotales = usuariosTotales;
    }

}

File: Python/src/gestorAplicacion/usuario/Usuario.py
# Autores:
#   MATEO ÁLVAREZ MURILLO
#   EFRAÍN GÓMEZ RAMÍREZ
#   LIBARDO JOSÉ NAVARRO PEDROZO
#   ANA SOFÍA GÓMEZ ZAPATA
#   SEBASTIÁN OCAMPO GALVIS

from abc import ABC, abstractmethod
from gestorAplicacion.administracion.Materia import Materia


class Usuario(ABC):
    _usuariosTotales = []

    def __init__(self, id, nombre, facultad, pw="0000"):
        self._id = id
        self._nombre = nombre
        self._pw = str(pw)
        self._facultad = facultad
        Usuario._usuariosTotales.append(self)


    @abstractmethod
    def __str__(self):
        pass

    # Por si alguien utiliza el otro constructor que estaban en Java

    # def __init__(self, id, nombre, facultad):
    #     self._id = id
    #     self._nombre = nombre
    #     self._facultad = facultad
    #     Usuario._usuariosTotales.append(self)
    @classmethod
    def mostrarUsuarios(cls) -> str:
        retorno = ""
        i = 1
        for usuario in cls.getUsuariosTotales():
            retorno += (
                str(i) + ". " + usuario._nombre + ", id: " + str(usuario._id) + "\n"
            )
            i += 1
        return retorno

    def comprobacionFacultad(self, usuario) -> bool:
        facultad1 = self.getFacultad().lower()
        facultad2 = usuario.getFacultad().lower()
        return facultad1 == facultad2

    def desmatricularDelSistema(usuario) -> None:
        usuariosTotales = Usuario.getUsuariosTotales()
        for user in usuariosTotales:
            if usuario == user:
                usuariosTotales.remove(usuario)
                break

    def eliminar_materia(materia) -> None:
        Materia.getMateriasTotales().remove(materia)

    def agregar_materia(
        nombre, codigo, descripcion, creditos, facultad, prerrequisitos
    ) -> None:
        nuevaMateria = Materia(
            nombre, codigo, descripcion, creditos, facultad, prerrequisitos
        )
        Materia.getMateriasTotales().append(nuevaMateria)

    def getTipo(self):
        return self._tipo

    def setTipo(self, tipo):
        self._tipo = tipo

    def getId(self):
        return self._id

    def setId(self, id):
        self._id = id

    def getNombre(self):
        return self._nombre

    def setNombre(self, nombre):
        self._nombre = nombre

    def getPw(self):
        return self._pw

    def setPw(self, pw):
        self._pw = pw

    def getFacultad(self):
        return self._facultad

    def setFacultad(self, facultad):
        self._facultad = facultad

    @classmethod
    def getUsuariosTotales(cls):
        return cls._usuariosTotales

    @classmethod
    def setUsuariosTotales(cls, usuariosTotales):
        cls._usuariosTotales = usuariosTotales
